R/boilerplates.R

```diff
@@ -0,0 +1,129 @@
+#'
+#' @return A function
+#' @export
+#'
+#' @examples
+#' # For testing purposes, a geom that returns grobs
+#' GeomTest <- ggproto(
+#'   "GeomTest", Geom,
+#'   draw_group = function(..., grob = grid::pointsGrob()) {
+#'     return(grob)
+#'   }
+#' )
+#' # Creating a constructor
+#' geom_test <- boilerplate(GeomTest)
+#'
+#' # Note that `grob` is automatically an argument to the function
+#' names(formals(geom_test))
+#'
+#' # Use in a plot
+#' set.seed(1234)
+#' p <- ggplot(mtcars, aes(disp, mpg))
+#' p + geom_test()
+#' p + geom_test(grob = grid::circleGrob())
+boilerplate <- function(x, ...) {
+  UseMethod("boilerplate")
+}
+
+#' @export
+boilerplate.Geom <- function(x, ..., checks, env = caller_env()) {
+
+  # Check that we can independently find the geom
+  geom <- gsub("^geom_", "", snake_class(x))
+  check_subclass(geom, "Geom", env = env)
+
+  # Split additional arguments into required and extra ones
+  args <- enexprs(...)
+  fixed_fmls_names <- c("mapping", "data", "stat", "position", "...",
+                        "na.rm", "show.legend", "inherit.aes")
+  extra_args <- setdiff(names(args), fixed_fmls_names)
+  if ("geom" %in% extra_args) {
+    cli::cli_abort("{.arg geom} is a reserved argument.")
+  }
+
+  # Fill in values for parameters from draw functions
+  known_params <-
+    unique(c(names(args), fixed_fmls_names, "flipped_aes", x$aesthetics()))
+  missing_params <- setdiff(x$parameters(), known_params)
+  if (length(missing_params) > 0) {
+    draw_args <- ggproto_formals(x$draw_panel)
+    if ("..." %in% names(draw_args)) {
+      draw_args <- ggproto_formals(x$draw_group)
+    }
+    params <- intersect(missing_params, names(draw_args))
+    extra_args <- c(extra_args, params)
+    for (param in params) {
+      if (!identical(draw_args[[param]], quote(expr = ))) {
+        args[param] <- draw_args[param]
+      }
+    }
+    missing_params <- setdiff(missing_params, names(args))
+    if (length(missing_params) > 0) {
+      cli::cli_warn(
+        "In {.fn geom_{geom}}: please consider providing default values for: \\
+        {missing_params}."
+      )
+    }
+  }
+
+  # Build function formals
+  fmls <- list2(
+    mapping  = args$mapping,
+    data     = args$data,
+    stat     = args$stat %||% "identity",
+    position = args$position %||% "identity",
+    `...` = quote(expr = ),
+    !!!args[extra_args],
+    na.rm    = args$na.rm %||% FALSE,
+    show.legend = args$show.legend %||% NA,
+    inherit.aes = args$inherit.aes %||% TRUE
+  )
+
+  if (length(extra_args) > 0) {
+    extra_args <- paste0(
+      "\n      ", extra_args, " = ", extra_args, ",", collapse = ""
+    )
+  }
+
+  body <- paste0("
+  layer(
+    data = data,
+    mapping = mapping,
+    stat = stat,
+    geom = \"", geom, "\",
+    position = position,
+    show.legend = show.legend,
+    inherit.aes = inherit.aes,
+    params = list2(
+      na.rm = na.rm,",
+      extra_args, "
+      ...
+    )
+  )
+  ")
+  body <- str2lang(body)
```

Comment:
I would resolve the trivial `arg = arg` type arguments of `layer()` first:

```r
  # Resolve layer(...) arguments
  layer_args <- rlang::exprs(
    data, mapping, stat, position, show.legend, inherit.aes,
    .named = TRUE
  )
```

And then deal with `params` separately:

```r
  # Resolve layer(params = list2(...))
  layer_params_arg <- rlang::exprs(
    na.rm, !!!rlang::syms(extra_args),
    .named = TRUE
  )
  layer_params_arg <- c(layer_params_arg, quote(...))
  layer_args$params <- rlang::call2("list2", !!!layer_params_arg, .ns = "rlang")
```

Finally, put the arguments inside `layer()` call

```r
  # Complete layer() expression
  body <- rlang::call2("layer", !!!layer_args)
```

Reply:
Reminds me of `fortunes::fortune("parse()")` but for `str2lang()`